lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java

```diff
@@ -0,0 +1,211 @@
+        assertSanityOfResponse(
+                ImmutableList.of(BatchElement.of(
+                        NamespaceAndRequestParams.of(
+                                namespace, RequestParams.of(127, getCache(namespace), lockCleanupService)),
+                        new DisruptorFuture<>("test"))),
+                false);
+    }
+
+    @Test
+    public void updatesCacheWhileProcessingResponse() {
+        Namespace namespace = Namespace.of("Test" + UUID.randomUUID());
+        assertSanityOfResponse(
+                ImmutableList.of(BatchElement.of(
+                        NamespaceAndRequestParams.of(
+                                namespace,
+                                RequestParams.of(
+                                        PARTITIONED_TIMESTAMPS_LIMIT_PER_SERVER_CALL - 1,
+                                        getCache(namespace),
+                                        lockCleanupService)),
+                        new DisruptorFuture<>("test"))),
+                true);
+        verify(getCache(namespace)).processStartTransactionsUpdate(any(), any());
+    }
+
+    @Test
+    public void shouldFreeResourcesIfServerThrows() {
+        Namespace namespace = Namespace.of("Test" + UUID.randomUUID());
+
+        UUID requestorId = UUID.randomUUID();
+        ImmutableList<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>>
+                requests = ImmutableList.of(BatchElement.of(
+                NamespaceAndRequestParams.of(
+                        namespace,
+                        RequestParams.of(
+                                PARTITIONED_TIMESTAMPS_LIMIT_PER_SERVER_CALL * 5,
+                                getCache(namespace),
+                                lockCleanupService)),
+                new DisruptorFuture<>("test")));
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap =
+                getMultiClientStartTransactionsResponse(requests, requestorId);
+
+        SafeIllegalStateException exception = new SafeIllegalStateException("Something went wrong!");
+        when(timelockService.startTransactions(any())).thenReturn(responseMap).thenThrow(exception);
+
+        assertThatThrownBy(() -> processBatch(timelockService, requestorId, requests))
+                .isEqualTo(exception);
+        verify(lockCleanupService).refreshLockLeases(any());
+        verify(lockCleanupService).unlock(any());
+    }
+
+    private void assertSanityOfResponse(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>>
+                    requestsForClients,
+            boolean assertValues) {
+
+        UUID requestorId = UUID.randomUUID();
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap =
+                getMultiClientStartTransactionsResponse(requestsForClients, requestorId);
+        when(timelockService.startTransactions(any())).thenReturn(responseMap);
```

Comment:
Ah, I see. This probably should use `thenAnswer(...)` so that you can also validate the values on the `canServiceOneRequestWithMultipleServerRequests` test